core/server/worker/src/main/java/alluxio/worker/netty/DataServerBlockReadHandler.java

```diff
@@ -52,26 +64,44 @@
    * up any temporary state it may have accumulated.
    */
   private final class BlockReadRequestInternal extends ReadRequestInternal {
-    final BlockReader mBlockReader;
+    BlockReader mBlockReader;
+    final Protocol.OpenUfsBlockOptions mOpenUfsBlockOptions;
 
     /**
      * Creates an instance of {@link BlockReadRequestInternal}.
      *
      * @param request the block read request
      */
     BlockReadRequestInternal(Protocol.ReadRequest request) throws Exception {
-      super(request.getId(), request.getOffset(), request.getOffset() + request.getLength(),
-          request.getPacketSize(), request.getSessionId());
-      mBlockReader = mWorker.readBlockRemote(mSessionId, mId, request.getLockId());
-      mWorker.accessBlock(mSessionId, mId);
+      super(request.getBlockId(), request.getOffset(), request.getOffset() + request.getLength(),
+          request.getPacketSize());
 
-      ((FileChannel) mBlockReader.getChannel()).position(mStart);
+      if (request.hasOpenUfsBlockOptions()) {
+        mOpenUfsBlockOptions = request.getOpenUfsBlockOptions();
+      } else {
+        mOpenUfsBlockOptions = null;
+      }
+      // Note that we do not need to seek to offset since the block worker is created at the offset.
     }
 
     @Override
-    public void close() throws IOException {
+    public void close() {
       if (mBlockReader != null) {
-        mBlockReader.close();
+        try {
+          mBlockReader.close();
+        } catch (Exception e) {
+          LOG.warn("Failed to close block reader for block {} with error {}.", mId, e.getMessage());
+        }
+      }
+
+      try {
+        if (mBlockReader instanceof UnderFileSystemBlockReader) {
```

Comment:
Is there a cleaner way to do this without relying on the type of the `BlockReader`?